File: logistic_regression/logistic_regression.R
# Read Weather data file
setwd(getwd())
pathFile <- file.path("weather.csv")
weatherData <- read.csv(pathFile, stringsAsFactors = FALSE)

# Ignore dates (in order to prevent error when predicting)
# Error: factor Date has new levels
weatherData <- subset(weatherData, select = -c(Date))

# Omit rows without available values
weatherData <- na.omit(weatherData)

set.seed(123)
# Calculate random indexes and split into two groups. Train data will include 70% of data
randomIndexes <- sample(1:nrow(weatherData))
splitRate <- floor(0.7 * nrow(weatherData))
trainRate <- randomIndexes[1:splitRate]
# Split data into train and test
trainData <- weatherData[trainRate, ]
testData <- weatherData[-trainRate, ]

# Convert RainTomorrow into 0 and 1 instead of No and Yes
trainData$RainTomorrow <- ifelse(trainData$RainTomorrow == "Yes", 1, 0)


model <- glm( RainTomorrow ~., data = trainData, family = binomial)
# POSSIBLE ERROR: 
# Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : contrasts can be applied only to factors with 2 or more levels
# Why? There is a column with a unique value for all rows:
# https://www.statology.org/contrasts-applied-to-factors-with-2-or-more-levels/
# Investigate different values per col:
# sapply(lapply(trainData, unique), length)

# Test the model
probabilities <- predict(model, testData, type = "response")
predicted.classes <- ifelse(probabilities > 0.5, "Yes", "No")

# Sacamos la precisión del modelo de test
mean(predicted.classes == testData$RainTomorrow)

File: logistic_regression/logistic_regression.py
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Read Weather data file
weatherData = pd.read_csv('weather.csv')

# Ignore dates (in order to prevent error when predicting)
weatherData = weatherData.drop('Date', axis=1)

# Omit rows without available values
weatherData = weatherData.dropna()
# Reset indexes after removing rows
weatherData = weatherData.reset_index(drop=True)

# Convert RainToday and RainTomorrow into 0 and 1 instead of No and Yes
weatherData["RainToday"] = np.where(weatherData["RainToday"] == "Yes", 1, 0)
weatherData["RainTomorrow"] = np.where(weatherData["RainTomorrow"] == "Yes", 1, 0)

X = weatherData.iloc[:, :-1]
y = weatherData.iloc[:, -1]

np.random.seed(123)
# Calculate random indexes and split into two groups. Train data will include 70% of data
indexes = np.random.permutation(len(weatherData))
train_size = int(len(weatherData) * 0.7)
train_idx, test_idx = indexes[:train_size], indexes[train_size:]
# Split data into train and test
X_train, y_train = X.iloc[train_idx, :], y[train_idx]
X_test, y_test = X.iloc[test_idx, :], y[test_idx]

# Variables that are measured at different scales do not contribute equally to the model fitting & model learned function and might end up creating a bias
sc_x = StandardScaler()
X_train = sc_x.fit_transform(X_train)
X_test = sc_x.transform(X_test)

# Train the model
model = LogisticRegression()
model.fit(X_train, y_train)

# Test the model
y_pred = model.predict(X_test)

# Get model accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Model accuracy on test data:", accuracy)

# Draw it
# TODO: Draw it